docs(schema): add NuxtHooks interface documentation (#18756)
nuxt/nuxt: docs/3.api/4.advanced/1.hooks.md

Check the [schema source code](https://github.com/nuxt/nuxt/blob/main/packages/schema/src/types/hooks.ts#L53) for all available hooks.

> Hook                     | Arguments                  | Environment     | Description

Reviewer: I think we can lose the environment column as they all happen within the build server.

Later, we can refactor to perhaps list these in the normal order they are called, but I wouldn't want to block the PR for that.

-------------------------|----------------------------|-----------------|-------------
`kit:compatibility`      | `compatibility, issues`    | Server          | Allows extending compatibility checks.

> `ready`                  | `nuxt`                     | Server          | Called after nuxt initialization, when the nuxt instance is ready to work.

Reviewer: For consistency, let's capitalise `nuxt`, so `Nuxt`.

`close`                  | `nuxt`                     | Server          | Called when Nuxt instance is gracefully closing.

> `modules:before`         | -                          | Server          | Called during nuxt initialization, before install user modules.
> `modules:done`           | -                          | Server          | Called during nuxt initialization, after install user modules.

Reviewer: ```suggestion
`modules:before`         | -                          | Server          | Called during Nuxt initialization, before installing user modules.
`modules:done`           | -                          | Server          | Called during Nuxt initialization, after installing user modules.
```

> `app:resolve`            | `app`                      | Server          | Called after resolve the `app` instance.

Reviewer: ```suggestion
`app:resolve`            | `app`                      | Server          | Called after resolving the `app` instance.
```

> `app:templates`          | `app`                      | Server          | Called during `NuxtApp` generation, after configure user templates.

Reviewer: ```suggestion
`app:templates`          | `app`                      | Server          | Called during `NuxtApp` generation, to allow customizing, modifying or adding new files to the build directory (either virtually or to written to `.nuxt`).
```

> `app:templatesGenerated` | `app`                      | Server          | Called after templates are compiled into virtual file system (vfs).

Reviewer: ```suggestion
`app:templatesGenerated` | `app`                      | Server          | Called after templates are compiled into the [virtual file system](https://nuxt.com/docs/guide/directory-structure/nuxt#virtual-file-system) (vfs).
```

`build:before`           | -                          | Server          | Called before Nuxt bundle builder.
`build:done`             | -                          | Server          | Called after Nuxt bundle builder is complete.

> `build:manifest`         | `manifest`                 | Server          | Called during the manifest build by Vite and Webpack.

Reviewer: ```suggestion
`build:manifest`         | `manifest`                 | Server          | Called during the manifest build by Vite and Webpack. This allows customizing the manifest that Nitro will use to render `<script>` and `<link>` tags in the final HTML.
```

> `builder:generateApp`    | `options`                  | Server          | Called before generate the app.

Reviewer: ```suggestion
`builder:generateApp`    | `options`                  | Server          | Called before generating the app.
```

> `builder:watch`          | `event, path`              | Server          | Called on build time when option.dev is enabled.

Reviewer: ```suggestion
`builder:watch`          | `event, path`              | Server          | Called at build time in development when the watcher spots a change to a file or directory in the project.
```

`pages:extend`           | `pages`                    | Server          | Called after pages routes are resolved.

> `server:devHandler`      | `handler`                  | Server          | Called when the dev middleware is been registered on server.

Reviewer: ```suggestion
`server:devHandler`      | `handler`                  | Server          | Called when the dev middleware is being registered on the Nitro dev server.
```

> `imports:sources`        | `presets`                  | Server          | Called at setup allowing modules extending sources.
> `imports:extend`         | `imports`                  | Server          | Called at setup allowing modules extending imports.

Reviewer: ```suggestion
`imports:sources`        | `presets`                  | Server          | Called at setup allowing modules to extend sources.
`imports:extend`         | `imports`                  | Server          | Called at setup allowing modules to extend imports.
```

> `imports:context`        | `context`                  | Server          | Called when the Unimport context is created.

Reviewer: ```suggestion
`imports:context`        | `context`                  | Server          | Called when the [unimport](https://github.com/unjs/unimport) context is created.
```

`imports:dirs`           | `dirs`                     | Server          | Allows extending import directories.

> `components:dirs`        | `dirs`                     | Server          | Called at app:resolve allowing to extend the directories.

Reviewer: ```suggestion
`components:dirs`        | `dirs`                     | Server          | Called within `app:resolve` allowing to extend the directories that are scanned for auto-importable components.
```

`components:extend`      | `components`               | Server          | Allows extending new components.

> `nitro:config`           | `nitroConfig`              | Server          | Called at nitro init allowing to extend the configuration.
> `nitro:init`             | `nitro`                    | Server          | Called when nitro is exposed at Nuxt instance.

Reviewer: ```suggestion
`nitro:config`           | `nitroConfig`              | Server          | Called before initializing Nitro, allowing customization of Nitro's configuration.
`nitro:init`             | `nitro`                    | Server          | Called after Nitro is initialized, which allows registering Nitro hooks and interacting directly with Nitro.
```

> `nitro:build:before`     | `nitro`                    | Server          | Called before build the nitro instance.
> `prerender:routes`       | `ctx`                      | Server          | Allows extending the routes to be prerender.

Reviewer: ```suggestion
`nitro:build:before`     | `nitro`                    | Server          | Called before building the nitro instance.
`prerender:routes`       | `ctx`                      | Server          | Allows extending the routes to be prerendered.
```

`build:error`            | `error`                    | Server          | Called when an error occurs at build time.

> `prepare:types`          | `options`                  | Server          | Called when nuxi is writing types.

Reviewer: ```suggestion
`prepare:types`          | `options`                  | Server          | Called before nuxi writes `.nuxt/tsconfig.json` and `.nuxt/nuxt.d.ts`, allowing addition of custom references and declarations in `nuxt.d.ts`, or directly modifying the options in `tsconfig.json`
```

`listen`                 | `listenerServer, listener` | Server          | Called when the dev server is loading.

> `schema:extend`          | `schemas`                  | Server          | Allows to extend default schemas.
> `schema:resolved`        | `schema`                   | Server          | Allows to extend resolved schema.
> `schema:beforeWrite`     | `schema`                   | Server          | Called before write the given schema.

Reviewer: ```suggestion
`schema:extend`          | `schemas`                  | Server          | Allows extending default schemas.
`schema:resolved`        | `schema`                   | Server          | Allows extending resolved schema.
`schema:beforeWrite`     | `schema`                   | Server          | Called before writing the given schema.
```

`schema:written`         | -                          | Server          | Called after the schema is written.
`vite:extend`            | `viteBuildContext`         | Server          | Allows to extend Vite default context.
`vite:extendConfig`      | `viteInlineConfig, env`    | Server & Client | Allows to extend Vite default config.

> `vite:serverCreated`     | `viteServer, env`          | Server & Client | Called when the vite server is created.
> `vite:compiled`          | -                          | Server & Client | Called after vite server is compiled.
> `webpack:config`         | `webpackConfigs`           | Server          | Called before configure the webpack compiler.

Reviewer: ```suggestion
`vite:serverCreated`     | `viteServer, env`          | Server & Client | Called when the Vite server is created.
`vite:compiled`          | -                          | Server & Client | Called after Vite server is compiled.
`webpack:config`         | `webpackConfigs`           | Server          | Called before configuring the webpack compiler.
```
